### PR TITLE
OAuth for Logged-Out Users

### PR DESCRIPTION
**Pull request by VukPerisic**
- Bypassed middleware by removing check for req.user for OAuth requests (Google Calendar and Microsoft Outlook)
- Hardcoded user ID of admin account in OAuth controller logic

### CODE REVIEW

**Comment by shawnpatel** [comment]
please update the name and description so highlight what you changed

File: packages/cli/src/auth/auth.service.ts [hunk 1]
```diff
@@ -57,9 +58,17 @@ export class AuthService {
...
 		}
-
-		if (req.user) next();
-		else res.status(401).json({ status: 'error', message: 'Unauthorized' });
+		
+		if (req.body) {
+			if (Object.keys(req.body).length === 0) { //oauth flow sends a lot of {} request bodies before the main request (these normally have req.user but not in our case so adding this logic) 
+				next(); 
+			} else if ("type" in req.body && (req.body.type === 'googleCalendarOAuth2Api' || req.body.type === 'microsoftOutlookOAuth2Api')) {
+				next();
```
Comment (shawnpatel) [inline packages/cli/src/auth/auth.service.ts]: good start! as discussed, let's flag by request url instead of auth type so we don't have to change this when we use more credentials

File: packages/cli/src/controllers/oauth/oAuth2Credential.controller.ts [hunk 2]
```diff
@@ -190,6 +195,7 @@ export class OAuth2CredentialController extends AbstractOAuthController {
 			delete decryptedDataOriginal.csrfSecret;
 			await this.encryptAndSaveData(credential, decryptedDataOriginal);
 
+			// add logic for storing credential.id and credential.type in db
```
Comment (shawnpatel) [inline packages/cli/src/controllers/oauth/oAuth2Credential.controller.ts]: wouldn't we do that on our backend (datalake) instead of here?

**Review by shawnpatel — CHANGES_REQUESTED** [review]

File: packages/cli/src/auth/auth.service.ts [hunk 3]
```diff
@@ -58,8 +59,11 @@ export class AuthService {
 			}
 		}
 
-		if (req.user) next();
-		else res.status(401).json({ status: 'error', message: 'Unauthorized' });
+		if (Object.keys(req.body).length === 0 || req.user) {
```
Comment (shawnpatel) [inline packages/cli/src/auth/auth.service.ts]: can you add comments as to why you made changes
i.e why are you checking `Object.keys(req.body).length === 0` to bypass auth?

File: packages/cli/src/controllers/oauth/oAuth2Credential.controller.ts [hunk 4]
```diff
@@ -17,15 +17,17 @@ interface CsrfStateParam {
...
 export class OAuth2CredentialController extends AbstractOAuthController {
 	override oauthVersion = 2;
 
 	/** Get Authorization url */
 	@Get('/auth')
 	async getAuthUri(req: OAuthRequest.OAuth2Credential.Auth): Promise<string> {
-		const credential = await this.getCredential(req);
-		const additionalData = await this.getAdditionalData(req.user);
+		const credential = await this.getCredential(req, hardcodedID);
+		const additionalData = await this.getAdditionalData(req.user, hardcodedID);
```
Comment (shawnpatel) [inline packages/cli/src/controllers/oauth/oAuth2Credential.controller.ts]: instead of adding a new id param, can we just set `user.id` to `hardcodedID`?

File: packages/core/src/NodeExecuteFunctions.ts
Comment (shawnpatel) [inline packages/core/src/NodeExecuteFunctions.ts:1]: bunch of unnecessary white space things here

File: packages/nodes-base/credentials/GmailOAuth2Api.credentials.ts [hunk 5]
```diff
@@ -10,7 +10,7 @@ const scopes = [
 ];
 
 export class GmailOAuth2Api implements ICredentialType {
-	name = 'gmailOAuth2';
+	name = 'gmailOAuth2Api';
```
Comment (shawnpatel) [inline packages/nodes-base/credentials/GmailOAuth2Api.credentials.ts]: is this change needed? don't see it used anywhere

File: packages/nodes-base/nodes/Google/Gmail/GmailTrigger.node.ts [hunk 6]
```diff
@@ -32,16 +32,7 @@ export class GmailTrigger implements INodeType {
...
 		credentials: [
 			{
-				name: 'googleApi',
-				required: true,
-				displayOptions: {
-					show: {
-						authentication: ['serviceAccount'],
-					},
-				},
-			},
```
Comment (shawnpatel) [inline packages/nodes-base/nodes/Google/Gmail/GmailTrigger.node.ts]: why is this removed? can't service account and oauth live together even if we only use oauth?

**Review by shawnpatel — CHANGES_REQUESTED** [review]
please be mindful of the changes you are making. everything should have a purpose. there are some whitespace additions/deletions here. please revert those so it doesn't cause conflicts when rebase with the public n8n repo.

also some other questions inline

File: packages/cli/src/controllers/oauth/abstractOAuth.controller.ts [hunk 7]
```diff
@@ -39,7 +39,6 @@ export abstract class AbstractOAuthController {
 		req: OAuthRequest.OAuth2Credential.Auth,
 	): Promise<CredentialsEntity> {
 		const { id: credentialId } = req.query;
-
```
Comment (shawnpatel) [inline packages/cli/src/controllers/oauth/abstractOAuth.controller.ts]: nit: add back newline

File: packages/nodes-base/nodes/Google/Calendar/GoogleCalendarTrigger.node.ts [hunk 8]
```diff
@@ -31,6 +31,15 @@ export class GoogleCalendarTrigger implements INodeType {
...
 		credentials: [
+			{
+				name: 'googleOAuth2Api',
+				required: true,
+				displayOptions: {
+					show: {
+						authentication: ['oAuth2'],
+					},
+				},
+			},
```
Comment (shawnpatel) [inline packages/nodes-base/nodes/Google/Calendar/GoogleCalendarTrigger.node.ts:42]: do we need this for the gmail trigger also?

File: packages/cli/src/controllers/oauth/oAuth2Credential.controller.ts [hunk 9]
```diff
@@ -195,6 +203,25 @@ export class OAuth2CredentialController extends AbstractOAuthController {
 				credentialId: credential.id,
 			});
 
+			// set up and trigger HTTP request node using new credentials to fetch user email from Google API
+			//^ this can be done with a direct request later but in localhost for now so using n8n's more secure request node
+
+			// send user email, credential id, and credential type to datalake backend
+			/*
+			const requestConfig: AxiosRequestConfig = {
+				url: 'http://localhost:5678/store/credential',  <-- route to be created on datalake backend
```
Comment (shawnpatel) [inline packages/cli/src/controllers/oauth/oAuth2Credential.controller.ts]: the api runs on port `8000`

File: packages/nodes-base/nodes/Google/Calendar/GoogleCalendarTrigger.node.ts (hunk 8 above)
Comment (VukPerisic) [reply packages/nodes-base/nodes/Google/Calendar/GoogleCalendarTrigger.node.ts:42]: Yeah it's in GmailTrigger too, the file shows only a one-line change since the other stuff (required, displayOptions, etc.) was there for 'GmailOAuth2' which I just changed to GoogleOAuth2Api

**Review by shawnpatel — APPROVED** [review]
